### PR TITLE
[inductor] propagate shapes in CSEVariable

### LINKED ISSUE #149905: [Inductor] track block shape of intermediary variables
### 🚀 The feature, motivation and pitch

During codegen each we track the dtype and value range of each intermediary variable we emit in trition. See [CSEVariable](https://github.com/pytorch/pytorch/blob/23855391f1a17f7145885b5ef977547a70819505/torch/_inductor/codegen/common.py#L1669-L1680). 

Dtype was recently added in https://github.com/pytorch/pytorch/pull/136778 by @arui-meta and subsequently iterated on in PRs like https://github.com/pytorch/pytorch/pull/141495 and https://github.com/pytorch/pytorch/pull/140057.

While dtypes are a bit finicky to get right, shapes are very easy to track in triton. More or less each operator broadcasts its inputs, reductions remove reduction dims, and then there are a few remaining ops.

@kundaMwiza recently [had an use case of shapes in a pr](https://github.com/pytorch/pytorch/pull/148679#issue-2900758922) `Ideally the shape of the input would be an attribute of a TritonCSEVariable via shape propagation`. 

Similarly, I [ran into a bug in prologue fusion](https://github.com/pytorch/pytorch/pull/147008/files#diff-73b89475038a5b4705da805f1217783883fb90398ee1164995db392fc4a342c1R773-R775) where I now need to add possibly extraneous broadcasts because in particular cases of loading a constant index we return a different shape.

I'm sure other future changes will run into needing shapes, and after adding we'll discover other places in the codebase we can simplify.
 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152198

Fixes #149905

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -74,6 +74,8 @@
     # causes typing errors in subclasses (defined in other files).
     OpVarT = str
 
+    ShapeType = Optional[Sequence[Union[int, str]]]
```
Comment (rec) [inline torch/_inductor/codegen/common.py:77]: There's already a `ShapeType` in [`_prims_common`](https://github.com/pytorch/pytorch/blob/a5f2fd1017d4d58a3fe7367bf682f95f7bfa38e9/torch/_prims_common/__init__.py#L51) defined as `Union[torch.Size, list[int], tuple[int, ...]]` and this type is already in use in `_inductor`.

What does it mean if a `Shape` has strings? Does it make sense to have `str` and `int` in the same collection?

Also, perhaps/probably the base type shouldn't be `Optional`? I see there are also `Optional[ShapeType]` type annotations below...?

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1761,6 +1765,7 @@ def generate(
         write: bool = True,
         assignment: bool = True,
         dtype: Optional[torch.dtype] = None,
+        shape: Optional[ShapeType] = None,
```
Comment (rec) [inline torch/_inductor/codegen/common.py:1768]: Does it have to be `Optional` here, either? Couldn't the default be `()`?

It makes the code simpler if you're not always checking for `None`.

File: torch/_inductor/codegen/halide.py [hunk 3]
```diff
@@ -1355,15 +1361,20 @@ def maybe_tuple(x):
         return tuple(unpack_vars)
 
     def genfunc(
-        self, line, used_dims, *, bounds=ValueRanges.unknown()
+        self,
+        line,
```
Comment (rec) [inline torch/_inductor/codegen/halide.py:1365]: Maybe add some types here, at least for the new parameter?

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -745,8 +746,10 @@ def low_precision_fp_var(var: Union[CSEVariable, Any]) -> bool:
 
 
 class TritonCSEVariable(CSEVariable):
-    def __init__(self, name, bounds: ValueRanges[Any], dtype: torch.dtype) -> None:
-        super().__init__(name, bounds, dtype)
+    def __init__(
+        self, name, bounds: ValueRanges[Any], dtype: torch.dtype, shape=None
```
Comment (rec) [inline torch/_inductor/codegen/triton.py:750]: Perhaps `name: str`?

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -3117,7 +3179,10 @@ def cse_multiple(line, values, masks, dtypes):
             cache_keys = [f"{line}, {i}, {masks}" for i in range(n)]
             if all(self.cse.contains(cache_key) for cache_key in cache_keys):
                 return [self.cse.get(cache_key) for cache_key in cache_keys]
-            result_vars = [self.cse.newvar(dtype=_dtype) for _dtype in dtypes]
+            result_vars = [
+                self.cse.newvar(dtype=_dtype, shape=_value.shape)
```
Comment (rec) [inline torch/_inductor/codegen/triton.py:3183]: `_value` and `_dtype` don't leak out into the enclosing, so you could just name them `value` and `dtype` or for such a short expression even `v` and `d`.

```
result_vars = [
    self.cse.newvar(dtype=d, shape=v.shape) for d, v in zip(dtypes, values)
]
```

File: torch/_inductor/shape_propagation.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+@functools.lru_cache(None)
+def get_broadcasted_shape(a: ShapeType, b: ShapeType) -> ShapeType:
+    if a is None:
+        return b
+    if b is None:
+        return a
+    assert isinstance(a, Sequence)
+    assert isinstance(b, Sequence)
+    if len(a) > len(b):
+        return get_broadcasted_shape(a, tuple(list(b) + list(a[len(b) :])))
```
Comment (rec) [inline torch/_inductor/shape_propagation.py:35]: What about: `*b, *a[len(b):]` instead of `tuple(list(b) + list(a[len(b) :])`

**Review by rec — COMMENTED** [review]
Sorry, quick unsolicited review while waiting for something else, feel free to ignore!

🙂

File: torch/_inductor/codegen/common.py (hunk 2 above)
Comment (isuruf) [reply torch/_inductor/codegen/common.py:1768]: `()` means the variable is a scalar. `None` means we don't know.

File: torch/_inductor/codegen/triton.py [hunk 7]
```diff
@@ -746,11 +748,19 @@ def low_precision_fp_var(var: Union[CSEVariable, Any]) -> bool:
...
+        bounds: ValueRanges[Any],
+        dtype: torch.dtype,
+        shape: Optional[ShapeType] = None,
+    ) -> None:
+        super().__init__(name, bounds, dtype, shape=shape)
         # We'll use this to track which masks the variable needs when used for indirect indexing
         self.mask_vars: OrderedSet[str] = OrderedSet()
         assert dtype is not None, "TritonCSEVariable must have dtype"
+        # TODO: uncomment this and fix the few failures left
+        # assert shape is not None, "TritonCSEVariable must have shape"
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:843]: I guess this is answers my other questions.. follow up good w me!

File: torch/_inductor/codegen/triton.py [hunk 8]
```diff
@@ -2406,7 +2456,21 @@ def reduction_resize(self, value) -> str:
...
+        ndims = self.triton_tensor_ndim()
+        if ndims == 1:
+            return f"triton_helpers.promote_to_tensor({value})", shape
+
+        nreduce = self.num_reduction_dims
+        sizes = [":"] * (ndims - nreduce) + ["None"] * nreduce
+        new_shape = (
+            (*shape[: (ndims - nreduce)], *[1] * nreduce) if shape is not None else None
+        )
+        return f"{value}[{', '.join(sizes)}]", new_shape
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:2814]: nit: if this is changing codegen, maybe better to wait until we have shapes everywhere/testing ?

Comment (eellison) [reply torch/_inductor/codegen/triton.py:2814]: See, tests/assertions added here https://github.com/pytorch/pytorch/commit/fd553b9817cb145e3cfaa52a08a831b9edf2fec7, and in the opinfo tests

File: torch/_inductor/codegen/triton.py [hunk 9]
```diff
@@ -2839,21 +2914,47 @@ def _welford(self, buffer, mean, m2, weight, dim, dtype: torch.dtype):
...
         welford = f"triton_helpers.welford({mean}, {m2}, {weight}, {dim})"
-        welford_results = [str(self.cse.newvar(dtype=dtype)) for _ in range(3)]
-        buffer.writeline(f"{', '.join(welford_results)} = {welford}")
 
-        result_values = tuple(self.reduction_resize(value) for value in welford_results)
-        return result_values
+        def reduced_shape(shape):
+            result = list(shape)
+            del result[dim]
+            return tuple(result)
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: nit: `tuple(shape[0:dim] + shape[dim+1:])`

File: torch/_inductor/shape_propagation.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+from typing import Callable, Optional, Protocol, Union
+
+import sympy
+
+import torch
+
+from .virtualized import OpsValue, V
+
+
+ShapeType = Optional[Sequence[Union[int, str]]]
```
Comment (eellison) [inline torch/_inductor/shape_propagation.py]: should this type be defined once ?

File: torch/_inductor/shape_propagation.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                result_shape = tuple(shape)
+            else:
+                result_shape = get_broadcasted_shape(result_shape, tuple(shape))
+        elif isinstance(arg, (int, float)):
+            if result_shape is None:
+                result_shape = ()
+        elif isinstance(arg, torch.dtype):
+            continue
+        else:
+            return None
```
Comment (eellison) [inline torch/_inductor/shape_propagation.py]: nit: do we know what types these are ? maybe assert ?

File: torch/_inductor/shape_propagation.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    ) -> None:
+        return None
+
+    @staticmethod
+    def to_dtype(
+        value: ShapeVar,
+        dtype: torch.dtype,
+        src_dtype: Optional[torch.dtype] = None,
+        use_compute_types: bool = True,
+    ) -> ShapeType:
```
Comment (eellison) [inline torch/_inductor/shape_propagation.py]: just for my understanding - reason why we need this ? I would assume since single ShapeVar input, we would take that shape.

File: torch/_inductor/codegen/common.py [hunk 13]
```diff
@@ -2366,45 +2378,64 @@ def _default(self, name: str, args: tuple[Any, ...], kwargs: dict[str, Any]) ->
...
+
         if name == "masked" and backend == "triton":
             output_dtype = value.dtype
+            output_shape = value.shape
         elif name == "masked" and backend == "cpp":
             output_dtype = V.interpreter.current_node.meta.get(
                 OptimizationContext.key, None
             ).dtype
+            # TODO: fix me
+            output_shape = None
```
Comment (eellison) [inline torch/_inductor/codegen/common.py:2454]: +1 , cpp not needed in initial pr..

File: torch/_inductor/codegen/common.py [hunk 14]
```diff
@@ -1848,8 +1854,12 @@ def generate(
             assert isinstance(expr, str)
             cache_key = expr
         var = self.try_get(cache_key)
+        if shape is None and not assignment:
+            # since there's no assignment to a variable, use any shape here
```
Comment (eellison) [inline torch/_inductor/codegen/common.py:1915]: would it make sense to make the shape the assigned variables shape ?

File: torch/_inductor/codegen/common.py [hunk 15]
```diff
@@ -79,6 +80,8 @@
     # causes typing errors in subclasses (defined in other files).
     OpVarT = str
 
+    ShapeType = Optional[Sequence[Union[int, str]]]
```
Comment (eellison) [inline torch/_inductor/codegen/common.py]: Should we make this non-Optional ? Scalars have a shape of (). As with dtypes in codegen, its much more useful if we can 100% rely on these instead of being sometimes present.

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/codegen/common.py]: Yea, I could see the confusion.. Maybe `BlockShapeType` or `VarShapeType` ? up to you.

**Review by eellison — APPROVED** [review]
Looks great! Excited to get this landed. The one comment that I think should definitely be addressed is to not use this in codegen since it is not 100% comprehensive/needs testing. 

Excited to get this working, clean up some code, and also turn on lazy broadcasting (which was an issue with the native matmul codegen)

**Comment by nullplay** [comment]
Hi, very excited to see this feature landing — I was actually looking into tracking variable shapes and came across this PR.
Just to clarify, this looks like it's for generic `CSEVariable`, not specifically for `TritonCSEVariable`, right?

File: torch/_inductor/shape_propagation.py (hunk 12 above)
Comment (isuruf) [reply torch/_inductor/shape_propagation.py]: If there are non-Shapevar inputs, we return None to say that we are not sure about the shape. This is because there are operations where the shape might be implicitly in one of the inputs. (For eg: index_expr)

File: torch/_inductor/codegen/common.py (hunk 14 above)
Comment (isuruf) [reply torch/_inductor/codegen/common.py:1915]: We could if `expr` was a `CSEVariable`, but not in other cases.

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (isuruf) [reply torch/_inductor/codegen/common.py]: `BlockShapeType` sounds good to me

File: torch/_inductor/codegen/common.py (hunk 15 above)
Comment (isuruf) [reply torch/_inductor/codegen/common.py]: Eventually yes, but we don't have 100% coverage to make it so. (For eg: with cpp backend)

File: torch/_inductor/shape_propagation.py (hunk 11 above)
Comment (isuruf) [reply torch/_inductor/shape_propagation.py]: You mean raise an assertion error if the type is unknown?

Comment (eellison) [reply torch/_inductor/shape_propagation.py]: Yea, assert what the type might be, so that it's clearer reading code / defensive against other new, unhandled types

File: torch/_inductor/codegen/triton.py (hunk 8 above)
Comment (isuruf) [reply torch/_inductor/codegen/triton.py:2814]: This doesn't change codegen, just adds a new function that returns the shape as well as `reduction_resize` codegen.

**Comment by jeffdaily** [comment]
@isuruf FYI this PR needed a forward fix for ROCm MI300 tests.  See https://github.com/pytorch/pytorch/pull/161166.
